utils/ffmpegvideodecode/ffmpeg_video_dec.cpp

```diff
@@ -0,0 +1,866 @@
+        case rocDecVideoCodec_JPEG  : return AV_CODEC_ID_MJPEG;
+        case rocDecVideoCodec_AV1   : return AV_CODEC_ID_AV1;
+        default                     : return AV_CODEC_ID_NONE;
+    }
+}
+
+static inline float GetChromaWidthFactor(rocDecVideoSurfaceFormat surface_format) {
+    float factor = 0.5;
+    switch (surface_format) {
+    case rocDecVideoSurfaceFormat_NV12:
+    case rocDecVideoSurfaceFormat_P016:
+        factor = 1.0;
+        break;
+    case rocDecVideoSurfaceFormat_YUV444:
+    case rocDecVideoSurfaceFormat_YUV444_16Bit:
+        factor = 1.0;
+        break;
+    case rocDecVideoSurfaceFormat_YUV420:
+    case rocDecVideoSurfaceFormat_YUV420_16Bit:
+        factor = 0.5;
+        break;
+    }
+    return factor;
+};
+
+
+/**
+ * @brief helper function for inferring AVCodecID from rocDecVideoSurfaceFormat
+ * 
+ * @param rocdec_codec 
+ * @return AVCodecID 
+ */
+static inline rocDecVideoSurfaceFormat AVPixelFormat2rocDecVideoSurfaceFormat(AVPixelFormat av_pixel_format) {
+    switch (av_pixel_format) {
+        case AV_PIX_FMT_YUV420P : 
+        case AV_PIX_FMT_YUVJ420P : 
+            return rocDecVideoSurfaceFormat_YUV420;           // need to see if this is actually correct
+        case AV_PIX_FMT_YUV444P : 
+        case AV_PIX_FMT_YUVJ444P : 
+            return rocDecVideoSurfaceFormat_YUV444;
+        case AV_PIX_FMT_YUV420P10LE :
+        case AV_PIX_FMT_YUV420P12LE :
+            return rocDecVideoSurfaceFormat_YUV420_16Bit;
+        default :
+            std::cerr << "ERROR: " << av_get_pix_fmt_name(av_pixel_format) << " pixel_format is not supported!" << std::endl;          
+            return rocDecVideoSurfaceFormat_NV12;       // for sanity
+    }
+}
+
+FFMpegVideoDecoder::FFMpegVideoDecoder(int device_id, OutputSurfaceMemoryType out_mem_type, rocDecVideoCodec codec, bool force_zero_latency,
+              const Rect *p_crop_rect, bool extract_user_sei_Message, uint32_t disp_delay, int max_width, int max_height, uint32_t clk_rate) :
+              RocVideoDecoder(device_id, out_mem_type, codec, force_zero_latency, p_crop_rect, extract_user_sei_Message, disp_delay, max_width, max_height, clk_rate) {
+
+    if ((out_mem_type_ == OUT_SURFACE_MEM_DEV_INTERNAL) || (out_mem_type_ == OUT_SURFACE_MEM_NOT_MAPPED)) {
+        THROW("Output Memory Type is not supported");
+    }
+    if (rocdec_parser_) {
+        rocDecDestroyVideoParser(rocdec_parser_);       // need to do this here since it was already created in the base class
+        // create rocdec videoparser
+        RocdecParserParams parser_params = {};
+        parser_params.codec_type = codec_id_;
+        parser_params.max_num_decode_surfaces = 1; // let the parser to determine the decode buffer pool size
+        parser_params.clock_rate = clk_rate;
+        parser_params.max_display_delay = disp_delay_;
+        parser_params.user_data = this;
+        parser_params.pfn_sequence_callback = FFMpegVideoDecoder::FFMpegHandleVideoSequenceProc;
```

Comment:
can we call the functions directly since it's member functions of the same class. Do we need the scope?